test/Microsoft.ML.Tests/FeatureContributionTests.cs

```diff
@@ -0,0 +1,344 @@
+        }
+
+        [Fact]
+        public void TestSSGDBinary()
+        {
+            TestFeatureContribution(ML.BinaryClassification.Trainers.SymbolicStochasticGradientDescent(), GetSparseDataset(TaskType.BinaryClassification, 100), "SSGDBinary", 4);
+        }
+
+        [Fact]
+        public void TestGAMBinary()
+        {
+            TestFeatureContribution(ML.BinaryClassification.Trainers.GeneralizedAdditiveModels(), GetSparseDataset(TaskType.BinaryClassification, 100), "GAMBinary");
+        }
+
+        private void TestFeatureContribution(
+            ITrainerEstimator<ISingleFeaturePredictionTransformer<IPredictor>, IPredictor> trainer,
+            IDataView data,
+            string testFile,
+            int precision = 6)
+        {
+            // Train the model.
+            var model = trainer.Fit(data);
+
+            // Extract the predictor, check that it supports feature contribution.
+            var predictor = model.Model as ICalculateFeatureContribution;
+            Assert.NotNull(predictor);
+
+            // Calculate feature contributions.
+            var est = new FeatureContributionCalculatingEstimator(ML, predictor, "Features", numPositiveContributions: 3, numNegativeContributions: 0)
+                .Append(new FeatureContributionCalculatingEstimator(ML, predictor, "Features", numPositiveContributions: 0, numNegativeContributions: 3))
+                .Append(new FeatureContributionCalculatingEstimator(ML, predictor, "Features", numPositiveContributions: 1, numNegativeContributions: 1))
+                .Append(new FeatureContributionCalculatingEstimator(ML, predictor, "Features", numPositiveContributions: 1, numNegativeContributions: 1, normalize: false));
+
+            TestEstimatorCore(est, data);
+            // Verify output.
+            var outputPath = GetOutputPath("FeatureContribution", testFile + ".tsv");
+            using (var ch = Env.Start("save"))
+            {
+                var saver = new TextSaver(ML, new TextSaver.Arguments { Silent = true, OutputHeader = false });
+                var savedData = ML.Data.TakeRows(est.Fit(data).Transform(data), 4);
+                using (var fs = File.Create(outputPath))
+                    DataSaverUtils.SaveDataView(ch, saver, savedData, fs, keepHidden: true);
+            }
+            CheckEquality("FeatureContribution", testFile + ".tsv", digitsOfPrecision: precision);
+            Done();
+        }
+
+        /// <summary>
+        /// Features: x1, x2vBuff(sparce vector), x3. 
+        /// y = 10x1 + 10x2vBuff + 30x3 + e.
+        /// Within xBuff feature  2nd slot will be sparse most of the time.
+        /// 2nd slot of xBuff has the least importance: Evaluation metrics do not change a lot when this slot is permuted.
+        /// x3 has the biggest importance.
+        /// </summary>
+        private IDataView GetSparseDataset(TaskType task = TaskType.Regression, int numberOfInstances = 1000)
```

Comment:
Can we put these into a common static library where it can be accessed from all the places that use it, rather than copy it here?